README.md

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status](https://travis-ci.org/czim/laravel-json-context-logging.svg?branch=master)](https://travis-ci.org/czim/laravel-json-context-logging)
[![Coverage Status](https://coveralls.io/repos/github/czim/laravel-json-context-logging/badge.svg?branch=master)](https://coveralls.io/github/czim/laravel-json-context-logging?branch=master)


# JSON Context Logging for Laravel

Wrapper package to make use of [czim/monolog-json-context](https://github.com/czim/monolog-json-context) in Laravel applications.

The aim of this package is to easily set up (separate) logging to be easily grokkable by Logstash.


## Version Compatibility

| Laravel       | Package |
|:--------------|:--------|
| 5.8 and below | 0.9     |
| 6.0 and up    | 1.0     |

## Changelog

[View the changelog](CHANGELOG.md).


## Installation

Via Composer:

``` bash
$ composer require czim/laravel-json-context-logging
```

Publish the configuration file:

``` bash
php artisan vendor:publish --provider="Czim\LaravelJsonContextLogging\Providers\JsonContextLoggingServiceProvider"
```



## Credits

- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/czim/laravel-json-context-logging.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/czim/laravel-json-context-logging.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/czim/laravel-json-context-logging
[link-downloads]: https://packagist.org/packages/czim/laravel-json-context-logging
[link-author]: https://github.com/czim
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

### [3.0.0] - 2023-04-07

Updated for monolog 3.0+.
For Laravel 10 and up.

### [2.0.0] - 2023-02-01

Updated to new psr/log interface (3.0.0+).
For Laravel 9.

### [1.0.3] - 2020-08-28

Added config setting `directories.file_chmod` to set file permissions for newly created log files.
Fixed an issue with the config value for the application name (now only uses `env()`).

[1.0.3]: https://github.com/czim/laravel-json-context-logging/compare/1.0.1...1.0.3


src/Loggers/AbstractLogger.php
<?php

declare(strict_types=1);

namespace Czim\LaravelJsonContextLogging\Loggers;

use Illuminate\Support\Facades\File;
use Psr\Log\LoggerInterface;
use Stringable;

abstract class AbstractLogger implements LoggerInterface
{
    protected const LOG_FILENAME  = 'laravel.log';
    protected const LOGS_SUB_PATH = false;

    public function __construct(protected LoggerInterface $logger)
    {
    }


    /**
     * {@inheritDoc}
     */
    public function emergency(string|Stringable $message, array $context = []): void
    {
        $this->logger->emergency($message, $context);
    }

    /**
     * {@inheritDoc}
     */
    public function alert(string|Stringable $message, array $context = []): void
    {
        $this->logger->alert($message, $context);
    }

    /**
     * {@inheritDoc}
     */
    public function critical(string|Stringable $message, array $context = []): void
    {
        $this->logger->critical($message, $context);
    }

    /**
     * {@inheritDoc}
     */
    public function error(string|Stringable $message, array $context = []): void
    {
        $this->logger->error($message, $context);
    }

    /**
     * {@inheritDoc}
     */
    public function warning(string|Stringable $message, array $context = []): void
    {
        $this->logger->warning($message, $context);
    }

    /**
     * {@inheritDoc}
     */
    public function notice(string|Stringable $message, array $context = []): void
    {
        $this->logger->notice($message, $context);
    }

    /**
     * {@inheritDoc}
     */
    public function info(string|Stringable $message, array $context = []): void
    {
        $this->logger->info($message, $context);
    }

    /**
     * {@inheritDoc}
     */
    public function debug(string|Stringable $message, array $context = []): void
    {
        $this->logger->debug($message, $context);
    }

    /**
     * {@inheritDoc}
     */
    public function log($level, string|Stringable $message, array $context = []): void
    {
        $this->logger->log($level, $message, $context);
    }


    protected function prepareLogPath(): string
    {
        $directory = $this->getDirectory();
        $path      = rtrim($directory, DIRECTORY_SEPARATOR) . DIRECTORY_SEPARATOR . $this->getFileName();

        $this->makeSureDirectoryExists($directory);

        return $path;
    }

    protected function getDirectory(): string
    {
        return storage_path('logs' . (static::LOGS_SUB_PATH ? '/' . static::LOGS_SUB_PATH : null));
    }

    protected function getFileName(): string
    {
        return static::LOG_FILENAME;
    }

    protected function makeSureDirectoryExists(string $directory): void
    {
        if (! is_dir($directory)) {
            File::makeDirectory($directory, $this->directoryPermissions(), true, true);
        }
    }

    protected function directoryPermissions(): int
    {
        return config('json-context-logging.directories.chmod', 0755);
    }
}


src/Factories/LoggerFactory.php
<?php

declare(strict_types=1);

namespace Czim\LaravelJsonContextLogging\Factories;

use Czim\LaravelJsonContextLogging\Contracts\LoggerFactoryInterface;
use Czim\MonologJsonContext\Formatters\JsonContextFormatter;
use Czim\MonologJsonContext\Formatters\PureJsonContextFormatter;
use Illuminate\Contracts\Config\Repository as ConfigRepository;
use Illuminate\Support\Arr;
use Illuminate\Support\Facades\File;
use Monolog\Formatter\FormatterInterface;
use Monolog\Handler\HandlerInterface;
use Monolog\Handler\RotatingFileHandler;
use Monolog\Handler\StreamHandler;
use Monolog\Logger;
use Psr\Log\LoggerInterface;
use RuntimeException;
use Throwable;

class LoggerFactory implements LoggerFactoryInterface
{
    public function __construct(protected ConfigRepository $config)
    {
    }

    public function make(string $channel): LoggerInterface
    {
        $path     = $this->getConfig($channel, 'path');
        $file     = $this->getConfig($channel, 'file');
        $fullPath = rtrim($path, DIRECTORY_SEPARATOR) . DIRECTORY_SEPARATOR . $file;


        if ($this->shouldMakeDirectory()) {
            $this->makeDirectory($path);
        }


        $handlerClass      = $this->getConfig($channel, 'handler.class');
        $handlerParameters = $this->getConfig($channel, 'handler.parameters');

        $handler = $this->makeHandler($handlerClass, $fullPath, $handlerParameters);


        $formatter       = $this->getConfig($channel, 'formatter.type');
        $dateFormat      = $this->getConfig($channel, 'formatter.date_format');
        $application     = $this->getConfig($channel, 'context.application');
        $defaultCategory = $this->getConfig($channel, 'context.category');

        $handler->setFormatter(
            $this->makeFormatter($formatter, $dateFormat, $application, $defaultCategory)
        );


        return new Logger($channel, [$handler]);
    }


    protected function shouldMakeDirectory(): bool
    {
        return (bool) $this->config->get('json-context-logging.directories.make_if_not_exists');
    }

    protected function directoryRights(): int
    {
        return $this->config->get('json-context-logging.directories.chmod', 0755);
    }

    protected function logFileRights(): int
    {
        return $this->config->get('json-context-logging.directories.file_chmod', 0644);
    }

    protected function makeDirectory(string $path): void
    {
        $directory = pathinfo($path, PATHINFO_DIRNAME);

        if (File::isDirectory($directory)) {
            return;
        }

        File::makeDirectory($directory, $this->directoryRights(), true);
    }

    /**
     * @param string               $class
     * @param string|null          $path
     * @param array<string, mixed> $parameters
     * @return HandlerInterface
     */
    protected function makeHandler(string $class, ?string $path = null, array $parameters = []): HandlerInterface
    {
        $level = Arr::get($parameters, 'level', Logger::DEBUG);

        switch ($class) {
            case StreamHandler::class:
                try {
                    return new StreamHandler($path, $level, true, $this->logFileRights());
                } catch (Throwable $exception) {
                    throw new RuntimeException($exception->getMessage(), $exception->getCode(), $exception);
                }


            case RotatingFileHandler::class:
                return new RotatingFileHandler(
                    $path,
                    Arr::get($parameters, 'max_files', 0),
                    $level,
                    true,
                    $this->logFileRights()
                );

            default:
                throw new RuntimeException("No support for '{$class}' handler yet.");
        }
    }

    protected function makeFormatter(
        ?string $type,
        ?string $dateFormat,
        ?string $application,
        ?string $defaultCategory,
    ): FormatterInterface {
        return match ($type) {
            'pure'  => new PureJsonContextFormatter($dateFormat, $application, $defaultCategory),
            default => new JsonContextFormatter($dateFormat, $application, $defaultCategory),
        };
    }

    protected function getConfig(string $channel, string $key, mixed $default = null): mixed
    {
        $channelConfig = $this->config->get('json-context-logging.channels.' . $channel);

        if (! is_array($channelConfig) || ! Arr::has($channelConfig, $key)) {
            return $this->getDefaultConfig($key, $default);
        }

        return Arr::get($channelConfig, $key, $default);
    }

    protected function getDefaultConfig(string $key, mixed $default = null): mixed
    {
        return $this->config->get('json-context-logging.default.' . $key, $default);
    }
}


src/Providers/JsonContextLoggingServiceProvider.php
<?php

declare(strict_types=1);

namespace Czim\LaravelJsonContextLogging\Providers;

use Czim\LaravelJsonContextLogging\Contracts\LoggerFactoryInterface;
use Czim\LaravelJsonContextLogging\Factories\LoggerFactory;
use Illuminate\Support\ServiceProvider;

class JsonContextLoggingServiceProvider extends ServiceProvider
{
    public function boot(): void
    {
        $this->bootConfig();
    }

    public function register(): void
    {
        $this->registerConfig();
        $this->registerInterfaceBindings();
    }


    protected function registerConfig(): void
    {
        $this->mergeConfigFrom($this->configPath(), 'json-context-logging');
    }

    protected function registerInterfaceBindings(): void
    {
        $this->app->singleton(LoggerFactoryInterface::class, LoggerFactory::class);
    }

    protected function bootConfig(): void
    {
        $this->publishes([
            $this->configPath() => config_path('json-context-logging.php'),
        ]);
    }

    protected function configPath(): string
    {
        return realpath(dirname(__DIR__) . '/../config/json-context-logging.php');
    }
}


config/json-context-logging.php
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Default
    |--------------------------------------------------------------------------
    |
    | Default logging settings.
    |
    */

    'default' => [

        // The path to write logs to
        'path' => storage_path('logs'),

        // The filename for the default log
        'file' => 'context.log',


        'handler' => [
            // The class name for the handler to use
            'class'      => \Monolog\Handler\RotatingFileHandler::class,
            // Custom parameters for constructor
            'parameters' => [
                'max_files' => null,
                'level'     => \Monolog\Level::Debug->value,
            ],
        ],

        'formatter' => [
            // Available: 'standard' (= null); 'pure'
            'type'        => null,
            // Data format to pass to formatter (default = 'Y-m-d H:i:s')
            'date_format' => null,
        ],

        'context' => [
            // A short string identifying this application
            'application' => env('APP_NAME', env('APP_ENV')),
            // The default category name (any custom string) for log entries
            'category'    => null,
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Channels
    |--------------------------------------------------------------------------
    |
    | You can configure different logging paths and handlers per channel here.
    |
    | All keys are optional. Any not present will take their values from the
    | defaults configured above.
    |
    */

    'channels' => [

        'testing' => [
            'path'      => storage_path('logs/testing'),
            'file'      => 'testing.log',
            'handler'   => [
                'parameters' => [
                    'max_files' => 7,
                ],
            ],
            'formatter' => [
            ],
            'context' => [
                'category' => 'testing.default',
            ],
        ],

    ],

    /*
    |--------------------------------------------------------------------------
    | Directories
    |--------------------------------------------------------------------------
    */

    'directories' => [

        // Silently create a new directory for the logs if it does not yet exist.
        'make_if_not_exists' => true,
        // The flags to set for a newly created log dir.
        'chmod'              => 0755,
        'file_chmod'         => 0644,
    ],

];
